firebase-perf/src/test/java/com/google/firebase/perf/util/TimerTest.java

```diff
@@ -27,10 +28,39 @@
 /** Unit tests for {@link Timer}. */
 @RunWith(RobolectricTestRunner.class)
 public class TimerTest {
+  /**
+   * Margin of error for time comparisons, because time is sensitive, wall-clock especially. 100
+   * milliseconds is chosen because 1000 is the difference between metric units, and we want tests
+   * to catch errors like extra or forgotten unit conversion.
+   */
+  private static final long epsilon = TimeUnit.MILLISECONDS.toMicros(100);
 
   @Before
   public void setUp() {}
 
+  @Test
+  public void ofElapsedRealtime_extrapolatesWallClock() {
+    Timer processStart = Timer.ofElapsedRealtime(Process.getStartElapsedRealtime());
+    Timer reference = new Timer();
+
+    assertThat(
+            approximatelyEqual(
```

Comment:
Can you do strict equal?

Reply:
no because I observe an 8 to 10 milliseconds difference between them

Reply:
Can you check if you can leverage ShadowSystemClock to change the system time? 

http://robolectric.org/extending/
http://robolectric.org/javadoc/4.4/org/robolectric/shadows/ShadowSystemClock.html

Note: the api changes quite drastically between versions, so you might need to check out whats available for the robolectric version we are using. You should check out the tests to see how Robolectric advances the system time.

Eg. https://github.com/robolectric/robolectric/blob/0aaa58831c55599180d6d803aed546620c07ab11/robolectric/src/test/java/org/robolectric/shadows/ShadowLegacySystemClockTest.java

Reply:
TIL, this is really cool! 

Unfortunately although `SystemClock` is shadowed, `System` is not, even though `ShadowSystem` exists in Robolectric's repo:
https://github.com/robolectric/robolectric/blob/0aaa58831c55599180d6d803aed546620c07ab11/shadows/framework/src/main/java/org/robolectric/shadows/ShadowSystem.java#L6

But it's missing the `@Implementation` annotation so it's not actually working as a shadow, I don't know why they do this. This mean unfortunately for this *specific `ofElapsedRealtime_extrapolatesWallTime` implementation* it won't hep. 

However I still learned a lot reading Robolectric's `ShadowSystemClock` implementation and learned that `SystemClock` is **actually paused** during tests, and I improved other tests with this new understanding. 

For `ofElapsedRealtime_extrapolatesWallTime` I found an alternative way to write this test avoiding approximate equality comparisons, PTAL @raymondlam @jeremyjiang-dev !

Reply:
Cool! Thanks for looking into it, too bad we can't do anything with System.currentTimeInMillis.